### PR TITLE
[typescript][axios] Handle boolean enum correctly

### LINKED ISSUE #9024: [BUG] typescript-axios - boolean enum is generated as string enum
##### Description
following schema
```json
    "schemas": {
      "BooleanEnum": {
        "enum": ["true", "false"],
        "type": "boolean"
      },
```
currently generates
```ts
export enum BooleanEnum {
    True = 'true',
    False = 'false'
}

```
This bug is affecting every typescript codegen, though a fix does not fix all of those

##### openapi-generator version

latest master

##### OpenAPI declaration file content or url

https://gist.github.com/haedaal/a771d6f3df337b624ae8a2c2d176c0db

##### Suggest a fix

It should generate this
```ts
export enum BooleanEnum {
    True = true,
    False = false
}
```
or this because of TS18033
```ts
export type BooleanEnum = true | false
```

### PR DESCRIPTION
**Pull request by haedaal**
closes #9024 

with following schema
```json
    "schemas": {
      "BooleanEnum": {
        "enum": ["true", "false"],
        "type": "boolean"
      },
```
https://gist.github.com/haedaal/a771d6f3df337b624ae8a2c2d176c0db

typescript-fetch currently generates
```ts
export enum BooleanEnum {
    True = 'true',
    False = 'false'
}
```

with this pr
```ts
export type BooleanEnum = true | false
```

by using 'type' rather than 'enum' we breaks consistency, but TS18033 does not allow this
```ts
export enum BooleanEnum {
    True = true, // Only numeric enums can have computed members, but this expression has type 'true' ...
    False = false
}
```

this pr has 2 changes.
1. by adding isBoolean, generate model correctly
2. fix typescript-axios modelEnum.mustache

accidentally typescript-angular is fixed without template changes, but many others need their own fixes on template.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

cc @TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01) @topce (2018/10) @akehir (2019/07) @petejohansonxo (2019/11) @amakhrov (2020/02)

### CODE REVIEW

**Review by tusharchutani — APPROVED** [review]
This looks good

**Comment by macjohnny** [comment]
What is the use case to have an enum with two boolean values?

**Comment by haedaal** [comment]
I think it's main use case is boolean literal type, not an enum with two boolean value (though I took the example above)

```ts
type FooResponse = FooSucessResponse | FooFailureResponse
type FooSuccessResponse = {
  success: true,
  body: {
    foo: string
  }
}
type FooFailureResponse = {
  success: false,
  body: {
    reason: string
  }
}
```

by handling boolean enum correctly, typescript can infer body type if value of success is given

**Comment by haedaal** [comment]
@tusharchutani any plan to merge it..?

**Comment by amakhrov** [comment]
Is it even a valid or reasonable schema in the first place?
I know it satisfies the validator - but the following one also does.
```
type: boolean
enum: [true, false, fuzzy]
```
So it tells us more about strictness of the validator, rather than correctness of the spec.

`type: boolean` already unambiguously indicates that the only two possible values are `true` and `false` (but not `"true"` or `"false"`!)

In this case we just need to decide whether we give priority to `type` attribute or `enum` attribute. In the former case, the generated code should not emit an additional type at all - just use `boolean` for the model property. In the latter case a string enum, as it's generated today, sounds correct.

**Comment by kenisteward** [comment]
This is true the original schema is not valid. If you want to use the enum
properly it should not be 'true' but just true

**Comment by kenisteward** [comment]
Technically the base validation is invalid due to the fact that the items
'true' and 'false' are not actually Boolean anyway.

**Comment by haedaal** [comment]
as a result of assigning string value to the value of boolean enum,
following pattern is required for now
```ts
if (response.success === 'true') { }
```

**Comment by haedaal** [comment]
@amakhrov does it make sense? I'm wondering what's blocking point.

**Comment by amakhrov** [comment]
@haedaal I still struggle to understand the use case, honestly. Why define a boolean value in the spec as an enum? Why are the enum values are defined as strings in the spec - even though you clearly want them to be boolean in the generated code.

**Comment by haedaal** [comment]
> @haedaal I still struggle to understand the use case, honestly. Why define a boolean value in the spec as an enum? Why are the enum values are defined as strings in the spec - even though you clearly want them to be boolean in the generated code.


@amakhrov 

1. Why define a boolean value in the spec as an enum?
```ts
type GetNextOrderResponse = {success: true, data: Order} | {success: false, data?: unknown} // or any other FailReason type for data in fail case
```
2. Why are the enum values are defined as strings in the spec - even though you clearly want them to be boolean in the generated code.
> that's the whole point of this pr. now codegen is generating string enum value rather than boolean enum value even though I clearly specify them as a boolean! this pr is a bugfix (if not a hotfix)

**Comment by haedaal** [comment]
> This is true the original schema is not valid. If you want to use the enum properly it should not be 'true' but just true
> […](#)

that's exactly what I thought is a problem, but if you try it, making schema like this
```json
    "schemas": {
      "BooleanEnum": {
        "enum": [true, false],
        "type": "boolean"
      },
    }
```
does not fix the wrong result.

it's because we check only for number type and unconditionally wrap it's value with double quote as you can see here.
And this pr add "boolean" type to the cases where you have to use value as it is.

https://github.com/OpenAPITools/openapi-generator/blob/11deb438290483410b34ddf8dc10a858dc215244/modules/swagger-codegen/src/main/java/io/swagger/codegen/DefaultCodegen.java#L249-L255

https://github.com/OpenAPITools/openapi-generator/blob/5d58afe5e79c425452716d48dc7f6c2bf39147d7/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java#L719-L723

**Comment by amakhrov** [comment]
Ok, I think now I see what you're trying to do (sorry, it took me a while :) ).

You don't actually need `BooleanEnum = true | false` (since it's redundant). You want smth like `TrueBooleanEnum = true` and `FalseBooleanEnum = false` - is that correct?

If so, I would rather try to avoid generating an `enum` in the first place. The ideal generated model would be exactly as in your example
```
{success: true, data: Order}
```
(no enum whatsoever)

However, I admit this would probably be a bigger change in the templates. So as a quick solution your PR makes total sense to me

**Review by amakhrov — APPROVED** [review]

**Comment by haedaal** [comment]
> You don't actually need `BooleanEnum = true | false` (since it's redundant). You want smth like `TrueBooleanEnum = true` and `FalseBooleanEnum = false` - is that correct?

Yes! Exactly! :)

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java [hunk 1]
```diff
@@ -64,7 +64,7 @@
     public String defaultValue;
     public String arrayModelType;
     public boolean isAlias; // Is this effectively an alias of another simple type
-    public boolean isString, isInteger, isLong, isNumber, isNumeric, isFloat, isDouble, isDate, isDateTime, isShort, isUnboundedInteger;
+    public boolean isString, isInteger, isLong, isNumber, isNumeric, isFloat, isDouble, isDate, isDateTime, isShort, isUnboundedInteger, isBoolean;
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java:67]: The following function also needs to be updated as well:

https://github.com/OpenAPITools/openapi-generator/blob/5d58afe5e79c425452716d48dc7f6c2bf39147d7/modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java#L764

https://github.com/OpenAPITools/openapi-generator/blob/5d58afe5e79c425452716d48dc7f6c2bf39147d7/modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java#L854

https://github.com/OpenAPITools/openapi-generator/blob/5d58afe5e79c425452716d48dc7f6c2bf39147d7/modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java#L872

I'll update these in another PR after merging this one.

**Comment by wing328** [comment]
cc @OpenAPITools/generator-core-team as the change impacts the codegen model class.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java (hunk 1 above)
Comment (wing328) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java:67]: Filed https://github.com/OpenAPITools/openapi-generator/pull/9802

**Pull request merged**